sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -402,6 +501,391 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
+
+    def acquire(sz: Long): Boolean = {
+      if (sz == 0) {
+        true
+      } else {
+        synchronized {
+          if (inFlight == 0 || sz + inFlight < maxBytesInFlight) {
+            inFlight += sz
+            true
+          } else {
+            false
+          }
+        }
+      }
+    }
+
+    def release(sz: Long): Unit = synchronized {
+      inFlight -= sz
+    }
+  }
+
+  class RapidsShuffleThreadedBlockIterator(
+      fetcherIterator: RapidsShuffleBlockFetcherIterator,
+      serializer: GpuColumnarBatchSerializer)
+    extends Iterator[(Any, Any)] with Arm {
+    private val queued = new LinkedBlockingQueue[(Any, Any)]
+    private val futures = new mutable.Queue[Future[Option[BlockState]]]()
+    private val serializerInstance = serializer.newInstance()
+    private var readBlockedTime: Long = 0L
+    private var fetchTime: Long = 0L
+    private var waitTime: Long = 0L
+    private val limiter = new BytesInFlightLimiter(maxBytesInFlight)
+    private val fallbackIter: Iterator[(Any, Any)] = if (numReaderThreads == 1) {
+      // this is the non-optimized case, where we add metrics to capture the blocked
+      // time and the deserialization time as part of the shuffle read time.
+      new Iterator[(Any, Any)]() {
+        private var currentIter: Iterator[(Any, Any)] = _
+        override def hasNext: Boolean = fetcherIterator.hasNext || (
+            currentIter != null && currentIter.hasNext)
+
+        override def next(): (Any, Any) = {
+          val fetchTimeStart = System.nanoTime()
+          readBlockedTime = 0
+          if (currentIter == null || !currentIter.hasNext) {
+            val readBlockedStart = System.nanoTime()
+            val (_, stream) = fetcherIterator.next()
+            readBlockedTime = System.nanoTime() - readBlockedStart
+            currentIter = serializerInstance.deserializeStream(stream).asKeyValueIterator
+          }
+          val res = currentIter.next()
+          fetchTime = System.nanoTime() - fetchTimeStart
+          res
+        }
+      }
+    } else {
+      null
+    }
+
+    override def hasNext: Boolean = {
+      if (fallbackIter != null) {
+        fallbackIter.hasNext
+      } else {
+        !pendingIts.isEmpty ||
+          fetcherIterator.hasNext || futures.nonEmpty || queued.size() > 0
+      }
+    }
+
+    case class BlockState(blockId: BlockId, batchIter: SerializedBatchIterator) {
+      private var _bytesDeserializedSoFar: Long = 0L
+      def getNextBatchSize: Long = batchIter.tryReadNextHeader().getOrElse(0L)
+      def addDeserializedBytes(deserBytes: Long): Unit = synchronized {
+        _bytesDeserializedSoFar += deserBytes
+      }
+
+      def bytesDeserializedSoFar: Long = synchronized {
+        _bytesDeserializedSoFar
+      }
+    }
+
+    private val pendingIts =
+      new HashedPriorityQueue[BlockState]((b1: BlockState, b2: BlockState) => {
+        java.lang.Long.compare(b1.bytesDeserializedSoFar, b2.bytesDeserializedSoFar)
```

Comment:
This looks like we're basing a priority queue on values that are moving without re-inserting the nodes in the queue when the values change?  Would also be good to comment why we need to order the queue this way

Reply:
yes that makes sense, I forget I need to reinsert the values. Overall, I am trying to complete that block that is holding onto most memory first, I'll add a comment.

Reply:
On second thought, the values are changing once a `BlockState` is removed, and then re-added later to the priority queue only if we need it. Comment still applies, however

Reply:
Ok I have replaced this with an ArrayBuffer since the priority doesn't make a lot of sense in this scenario. We may in the future look into prioritizing by smallest first, given that we are trying to fit into a limit of bytes, but I can see issues with either the priority case or the arrival order case.